src/lib/core/style/add-focus-classes.ts

```diff
@@ -0,0 +1,80 @@
+import {Directive, Injectable, Optional, SkipSelf} from '@angular/core';
+
+
+/** Singleton that allows all instances of CdkAddFocusClasses to share document event listeners. */
+@Injectable()
+export class CdkFocusCauseDetector {
+  /** Whether a keydown event has just occurred. */
+  get keydownOccurred() { return this._keydownOccurred; }
+  private _keydownOccurred = false;
+
+  get mousedownOccurred() { return this._mousedownOccurred; }
+  private _mousedownOccurred = false;
+
+  constructor() {
+    document.addEventListener('keydown', () => {
+      this._keydownOccurred = true;
+      setTimeout(() => this._keydownOccurred = false, 0);
```

Comment:
In Material 1 we only had a timeout for touch events (for the buffering), but for normal `mousedown` events we didn't need one.

We always keep track of the timestamp for the last interaction and then compare it based on the time difference.
> https://github.com/angular/material/blob/master/src/core/services/interaction/interaction.js#L144
> https://github.com/angular/material/blob/master/src/components/button/button.js#L169